Orange/widgets/visualize/owvenndiagram.py

```diff
@@ -151,69 +170,81 @@ def __init__(self):
     @Inputs.data
     @check_sql_input
     def setData(self, data, key=None):
-        self.error()
+        self.Error.too_much_inputs.clear()
         if not self._inputUpdate:
-            # Store hints only on the first setData call.
-            self._storeHints()
             self._inputUpdate = True
-
         if key in self.data:
             if data is None:
                 # Remove the input
-                self._remove(key)
+                # Clear possible warnings.
+                self.warning()
+                del self.data[key]
             else:
                 # Update existing item
-                self._update(key, data)
+                self.data[key] = self.data[key]._replace(name=data.name, table=data)
+
         elif data is not None:
             # TODO: Allow setting more them 5 inputs and let the user
             # select the 5 to display.
             if len(self.data) == 5:
-                self.error("Venn diagram accepts at most five datasets.")
+                self.Error.too_much_inputs()
                 return
             # Add a new input
-            self._add(key, data)
-
-    def handleNewSignals(self):
-        self._inputUpdate = False
-
-        # Check if all inputs are from the same domain.
-        domains = [input.table.domain for input in self.data.values()]
-        samedomain = all(domain_eq(d1, d2) for d1, d2 in pairwise(domains))
-
-        self.samedomain = samedomain
-
-        has_identifiers = all(source_attributes(input.table.domain)
-                              for input in self.data.values())
-        has_any_identifiers = any(source_attributes(input.table.domain)
+            self.data[key] = _InputData(key, data.name, data)
+        self._setInterAttributes()
+
+    def data_equality(self):
+        """ Checks if all input datasets have same ids. """
+        if not self.data.values():
+            return True
+        sets = []
+        for val in self.data.values():
+            sets.append(set(val.table.ids))
+        inter = reduce(set.intersection, sets)
+        if len(inter) == max(map(len, sets)):
+            return True
+        return False
+
+    def settings_compatible(self):
+        self.Error.domain_mismatch.clear()
+        self.Error.instances_mismatch.clear()
+        if not self.usecols:
+            domains = [input.table.domain for input in self.data.values()]
+            samedomain = all((d1 == d2) for d1, d2 in pairwise(domains))
+            if not samedomain:
+                self.vennwidget.clear()
+                self.Error.domain_mismatch()
+                self.itemsets = OrderedDict()
+                return False
```

Comment:
These four lines repeat again below. You can extract this into a function (including `return False`) and replace these lines with `return name_of_that_function()`.